### PR TITLE
Streamline `x fmt` and improve its output

### PR DESCRIPTION
**Pull request by nnethercote**
- Removes the ability to pass paths to `x fmt`, because it's complicated and not useful, and adds `--all`.
- Improves `x fmt` output.
- Improves `x fmt`'s internal code.

r? @GuillaumeGomez

### CODE REVIEW

File: src/bootstrap/src/core/build_steps/format.rs [hunk 1]
```diff
@@ -24,20 +25,23 @@ fn rustfmt(src: &Path, rustfmt: &Path, paths: &[PathBuf], check: bool) -> impl F
     cmd.args(paths);
     let cmd_debug = format!("{cmd:?}");
     let mut cmd = cmd.spawn().expect("running rustfmt");
-    // Poor man's async: return a closure that'll wait for rustfmt's completion.
+    // Poor man's async: return a closure that might for rustfmt's completion (depending on the
```
Comment (lqd) [inline src/bootstrap/src/core/build_steps/format.rs:28]: I assume this should read as the following?

```suggestion
    // Poor man's async: return a closure that might wait for rustfmt's completion (depending on the
```

Comment (nnethercote) [reply src/bootstrap/src/core/build_steps/format.rs]: Fixed, thanks!

File: src/bootstrap/src/core/build_steps/format.rs [hunk 2]
```diff
@@ -319,16 +279,33 @@ pub fn format(build: &Builder<'_>, check: bool, paths: &[PathBuf]) {
...
             if entry.file_type().map_or(false, |t| t.is_file()) {
+                formatted_paths_ref.lock().unwrap().push({
+                    // `into_path` produces an absolute path. Try to strip `cwd` to get a shorter
+                    // relative path.
+                    let mut path = entry.clone().into_path();
+                    if let Ok(cwd) = cwd {
+                        if let Ok(path2) = path.strip_prefix(cwd) {
+                            path = path2.to_path_buf();
+                        }
+                    }
```
Comment (GuillaumeGomez) [inline src/bootstrap/src/core/build_steps/format.rs:298]: Small nit:

```suggestion
                    if let Ok(path2) = cwd.and_then(|cwd| path.strip_prefix(cwd)) {
                        path = path2.to_path_buf();
                    }
```

**Review by GuillaumeGomez — COMMENTED** [review]
Just a small nit but otherwise looks great to me, thanks!

r=me once nit fixed.

File: src/bootstrap/src/core/build_steps/format.rs (hunk 2 above)
Comment (nnethercote) [reply src/bootstrap/src/core/build_steps/format.rs:298]: I'm looking forward to let-chains reaching stable :)

Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/format.rs:298]: Me too...

**Comment by ChrisDenton** [comment]
> Removes the ability to pass paths to x fmt, because it's complicated and not useful

I'll just add that I found this useful and was surprised when I rebased and was met with a terse and slightly ambiguous error instead.

Passing paths was especially useful for generated code when you only want to automatically format the generated file. But I also fairly often used `./x fmt library/std` to only format std and not the whole code base.

**Comment by nnethercote** [comment]
Can you give me more details about your use case, especially the generated code?

And with the `x fmt library/std` case, does running `x fmt --all` give a different outcome?

**Comment by ChrisDenton** [comment]
Specifically in my case std generates the Windows API bindings using `src/tools/generate-windows-sys`, which writes to `library/std/src/sys/pal/windows/c/windows_sys.rs` that then needs to be formatted according to std's formatting config.

> And with the `x fmt library/std` case, does running `x fmt --all` give a different outcome?

That seems to work. Though that is noticeably slower instead of being near enough instant.

**Comment by nnethercote** [comment]
`x fmt --all` is slower, to be sure. On my machine `x fmt` with very few changes takes about 2.2 seconds, and `x fmt --all` takes about 3.3 seconds. So bootstrap stuff before formatting takes more time than the actual formatting. But I do have a fast machine, so your results may be different. How often do you generate the windows bindings?

**Comment by ChrisDenton** [comment]
Not that often. And to be clear, I'm not saying it's a big deal but I wouldn't quite characterise it as useless either.

**Comment by nnethercote** [comment]
Thanks for the info. The overall calculus was:
- Path support is rarely needed, `--all` subsumes it and is only marginally slower.
- Path support is surprisingly complicated. It was completely broken for a while until #125637 fixed it again, and a number of non-formatted files slipped into the repository in this time. The `ignore` library has some surprising behaviours when you use combinations of explicitly excluded and explicitly included paths.
- So overall, it's not worth keeping.

I have filed #126051 to improve the error message given when `x fmt` is given a path. I agree the current one is not clear.

**Pull request merged**